### PR TITLE
18532 - AM Button Implementation

### PR DESCRIPTION
**Pull request by jamespaologarcia**
*Issue #:* /bcgov/entity#18532
Added implementation for amalgamate now button. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namerequest license (Apache 2.0).

### CODE REVIEW

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

File: src/components/new-request/search.vue
Comment (severinbeauvais) [inline src/components/new-request/search.vue:715]: I recommend changing lines 710-711 to an "else if" :)

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:207]: Line these up with the first * on line 202. (See method above for example.)

Also, on line 206, the definition should be "The legal type of the amalgamated business."

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:226]: - it's a draft "amalgamation application"
- I think there's always a business type, so remove "if applicable"
- on line 226, same fix as line 206

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:232]: Use `FilingTypes.AMALGAMATION` here. [ref])(https://github.com/bcgov/bcrs-shared-components/blob/main/src/enums/filing-types.ts)

Do you mind also fixing lines 144, 156 and 264?

**Review by severinbeauvais — COMMENTED** [review]
~I think you need to update something in App.vue as well, somewhere around here:~ _fixed now_
https://github.com/bcgov/namerequest/blob/4ca645bf1a15eae9cfe0cc58a4e98295f430f2d9/src/App.vue#L261

This is in case the user was initially not logged in and was redirected to log in, to resume the flow when they return to Namerequest UI.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:1]: I think you also need to update the method at line 137.

Karim, I can't remember why there are 2 flows for creating an IA...?

Comment (JazzarKarim) [reply src/mixins/nr-affiliation-mixin.ts:1]: One is for the "Incorporate Now" button in Numbered while the other is for the "Incorporate Now" button after submitting the NR and going to Manage my Name Request (affiliate NR to account and incorporate using that).

**Review by severinbeauvais — COMMENTED** [review]
~You also need to update existing-request-display.vue to handle your second "todo checkbox" in your ticket.~ _fixed now_

It should look like this, except "Amalgamate Now":

image

This ties in to my comment about updating getBusinessRequest().

If this doesn't fit in your current ticket's estimate, let me know and we'll create a new ticket for it.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

File: src/App.vue
Comment (JazzarKarim) [inline src/App.vue:130]: James, can you please alphabetize this import please? Ditto lines 157 and 179.

File: src/components/dialogs/index.ts
Comment (JazzarKarim) [inline src/components/dialogs/index.ts:10]: Same here please (alphabetize).

File: src/interfaces/new-request-interface.ts
Comment (JazzarKarim) [inline src/interfaces/new-request-interface.ts:39]: Here too.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

File: src/App.vue
Comment (JazzarKarim) [reply src/App.vue]: Fixed, thanks. Resolving.

File: src/components/dialogs/index.ts
Comment (JazzarKarim) [reply src/components/dialogs/index.ts]: Fixed, thanks. Resolving.

File: src/interfaces/new-request-interface.ts
Comment (JazzarKarim) [reply src/interfaces/new-request-interface.ts]: Fixed, thanks. Resolving.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

**Comment by JazzarKarim** [comment]
James, I'm pressing on the amalgamate now button after an NR has been approved, but I don't see anything happening. Shouldn't we be getting an HTTP 400 error for now at least since the BE hasn't been setup yet?
amalgamate now button doing nothing

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

**Comment by JazzarKarim** [comment]
> James, I'm pressing on the amalgamate now button after an NR has been approved, but I don't see anything happening. Shouldn't we be getting an HTTP 400 error for now at least since the BE hasn't been setup yet? amalgamate now button doing nothing

Fixed, thanks!

**Review by JazzarKarim — APPROVED** [review]
LGTM James! 👍 Please wait for at least other approval.

File: src/components/existing-request/nr-approved-gray-box.vue
Comment (severinbeauvais) [inline src/components/existing-request/nr-approved-gray-box.vue:190]: It looks like this is still used on line 208.

File: src/components/existing-request/nr-approved-gray-box.vue
Comment (severinbeauvais) [inline src/components/existing-request/nr-approved-gray-box.vue:58]: I'm not understanding why you need `&& !isSupportedAmalgamation`, since that appears to alreay be part of the logic for showOpenExternalIcon (line 221).

In fact, I think `isSupportedAmalgamation` isn't doing what you think it is -- in the code below, it's a function, but here you don't pass any parameters (or use brackets), so this code is actually testing for the presence of the function (not its result).

File: src/components/existing-request/nr-approved-gray-box.vue
Comment (severinbeauvais) [inline src/components/existing-request/nr-approved-gray-box.vue:262]: Eventually, delete the debugging statements.

By the way, did you know that you can debug this directly in VS Code?

File: src/components/existing-request/nr-approved-gray-box.vue
Comment (severinbeauvais) [reply src/components/existing-request/nr-approved-gray-box.vue:62]: This event is handled in existing-request-display.vue.

image

I think you should do the same for amalgamations.

If it helps, I think you don't need to the "Amalgamate Now ->" (with external arrow) button here. You only need to show "Amalgamate Now" if it's a supported entity type (same as incorporation and registration).

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [reply src/mixins/nr-affiliation-mixin.ts]: Thanks for the fix. Constants and enums are always better!

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:789]: I'm not sure why this is different from incorporateRegisterYourBusiness() above. It should look almost identical I think.

I think affiliateOrLogin() should end up calling amalgamateNow() -- whatever the incorporation flow does, this should do almost identically.

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:792]: Obsolete function comment.

File: src/components/existing-request/nr-approved-gray-box.vue
Comment (severinbeauvais) [inline src/components/existing-request/nr-approved-gray-box.vue:177]: Do you still need NrAffiliationMixin, or isAuthenticated below?

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:799]: Doesn't createAffiliation also handle the "amalgamate now" functionality?

Why do you need amalgamateNow() here but there is no corresponding incorporateNow() -- how doesn't "incorporate now" work and why aren't you doing the same thing?

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

File: src/components/existing-request/existing-request-display.vue
Comment (jamespaologarcia) [reply src/components/existing-request/existing-request-display.vue:798]: It's in the same file but it isn't called in createAffiliation. Should I handle the if statement there instead to call amalgamateNow?

Comment (severinbeauvais) [reply src/components/existing-request/existing-request-display.vue:798]: 1. Line 797 shows up as a code error when I look at it in VS Code. Please fix it.

2. Since line 797 isn't working, how were you able to test this? Please make a best effort to test things before asking reviewers to spend their time on it. Otherwise, ask for help, or call out in the PR that you're not sure if this is the way to do something or you aren't able to test or something.

3. If I work down through the code...
```
existing-name-request.vue::affiliateOrLogin()
  |
  +-> nr-afficreateAffiliation()
    |
    +-> createBusiness()
      |
      +-> getBusinessRequest()
```
... then I don't see any handling for an amalgamation business request.

Granted, there seems to be some existing code duplication between createBusiness() and createBusinessIA() but let's ignore that for now.

Just do the same thing as for incorporation applications (except update the business request for an amalgamation filing).

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:372]: It looks like you no longer need this.

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:790]: This is a duplicate of incorporateRegisterYourBusiness() above. Why not call the original one on line 317? In fact, why not use the existing event for amalgamations?

I recommend renaming the event and method names to indicate that they're used for incorporations, registrations and amalgamations.

**Comment by severinbeauvais** [comment]
James, please indicate here (or better, in the ticket), what testing you've done, including:
- entities supported by FF (create draft)
- entities not supported by FF (go to COLIN)
- while not logged in
- while already logged in
- numbered amalgamations (button on search page)
- named amalgamations (button on existing request display page)

Thanks.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:155]: This looks like it won't work as expected.

Line 143 says "if not a firm then create an IA". But amalgamations are not firms, so an IA would be created, which is incorrect.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:144]: This looks correct now - thanks.

If you're ever in here again -- perish the thought! -- use `NrRequestActionCodes.AMALGAMATE` instead of "AML". (This isn't worth a new commit on its own.)

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://namerequest-dev--pr-748-w80ga369.web.app

**Comment by severinbeauvais** [comment]
James, are you ready to merge this, or do you want to do a bit more testing first?

**Comment by severinbeauvais** [comment]
> James, are you ready to merge this, or do you want to do a bit more testing first?

@jamespaologarcia , I wrote this on Nov 29. What's your answer please?

**Comment by jamespaologarcia** [comment]
> > James, are you ready to merge this, or do you want to do a bit more testing first?
> 
> @jamespaologarcia , I wrote this on Nov 29. What's your answer please?

Hi,

I sent the PR link to Omid. I'm waiting for confirmation before I merge it.

**Comment by severinbeauvais** [comment]
> > James, are you ready to merge this, or do you want to do a bit more testing first?
> 
> @jamespaologarcia , I wrote this on Nov 29. What's your answer please?

James replied: "All good from my side. I tested the cases you pointed out to me."

**Review by severinbeauvais — APPROVED** [review]

**Pull request merged**
